dds/DCPS/transport/shmem/ShmemDataLink.cpp

```diff
@@ -89,12 +97,117 @@ ShmemDataLink::open(const std::string& peer_address)
             this, peer_address_.c_str()), 1);
 
+  assoc_resends_task_ = make_rch<SmPeriodicTask>(reactor_task_->interceptor(),
+    ref(*this), &ShmemDataLink::resend_association_msgs);
+  assoc_resends_task_->enable(false, config().association_resend_period());
+
   return true;
 }
 
+int ShmemDataLink::make_reservation(const GUID_t& remote_sub, const GUID_t& local_pub,
+  const TransportSendListener_wrch& send_listener, bool reliable)
+{
+  const int result = DataLink::make_reservation(remote_sub, local_pub, send_listener, reliable);
+  send_association_msg(local_pub, remote_sub);
+  return result;
+}
+
+int ShmemDataLink::make_reservation(const GUID_t& remote_pub, const GUID_t& local_sub,
+  const TransportReceiveListener_wrch& receive_listener, bool reliable)
+{
+  const int result = DataLink::make_reservation(remote_pub, local_sub, receive_listener, reliable);
+  send_association_msg(local_sub, remote_pub);
+  // Resend until we get a response.
+  {
+    ACE_GUARD_RETURN(ACE_Recursive_Thread_Mutex, g, mutex_, result);
+    assoc_resends_.insert(std::pair<GuidPair, unsigned>(GuidPair(remote_pub, local_sub),
+      config().association_resend_max_count()));
+  }
+  return result;
+}
+
+void
+ShmemDataLink::send_association_msg(const GUID_t& local, const GUID_t& remote)
+{
+  VDBG((LM_DEBUG, "(%P|%t) ShmemDataLink::send_association_msg from %C to %C\n",
+    LogGuid(local).c_str(), LogGuid(remote).c_str()));
+
+  DataSampleHeader header_data;
+  header_data.message_id_ = REQUEST_ACK;
+  header_data.byte_order_ = ACE_CDR_BYTE_ORDER;
+  header_data.message_length_ = guid_cdr_size;
+  header_data.sequence_ = -1;
+  header_data.publication_id_ = local;
+  header_data.publisher_id_ = remote;
+
+  Message_Block_Ptr message(
+    new ACE_Message_Block(header_data.get_max_serialized_size(),
+                          ACE_Message_Block::MB_DATA,
+                          0, //cont
+                          0, //data
+                          0, //allocator_strategy
+                          0, //locking_strategy
+                          ACE_DEFAULT_MESSAGE_BLOCK_PRIORITY,
+                          ACE_Time_Value::zero,
+                          ACE_Time_Value::max_time,
+                          0,
+                          0));
+
+  *message << header_data;
+  Serializer ser(message.get(), encoding_unaligned_native);
+  ser << remote;
+  send_strategy_->link_released(false);
+  TransportControlElement* send_element = new TransportControlElement(move(message));
+  this->send_i(send_element, false);
+}
+
+void ShmemDataLink::resend_association_msgs(const MonotonicTimePoint&)
+{
+  VDBG((LM_DEBUG, "(%P|%t) ShmemDataLink::resend_association_msgs\n"));
+
+  ACE_GUARD(ACE_Recursive_Thread_Mutex, g, mutex_);
+  for (AssocResends::iterator i = assoc_resends_.begin(); i != assoc_resends_.end();) {
+    send_association_msg(i->first.local, i->first.remote);
+    i->second--;
```

Comment:
```suggestion
    --i->second;
```
or combine with next line (the conditional for the `if`)